### ISSUE TITLE
Couldn't import movie NOT NULL

### ISSUE THREAD

**Issue by Firemad03** [Radarr/Radarr#3280]
Radarr version: 0.2.0.1291

Couldn't import movie /mnt/media/transmission/complete/movies-multi/Small Town Crime.mkv: constraint failed
NOT NULL constraint failed: MovieFiles.Quality

Debug Log

```
System.Data.SQLite.SQLiteException (0x80004005): constraint failed
NOT NULL constraint failed: MovieFiles.Quality
  at System.Data.SQLite.SQLite3.Reset (System.Data.SQLite.SQLiteStatement stmt) [0x00083] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLite3.Step (System.Data.SQLite.SQLiteStatement stmt) [0x0003c] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLiteDataReader.NextResult () [0x0016b] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLiteDataReader..ctor (System.Data.SQLite.SQLiteCommand cmd, System.Data.CommandBehavior behave) [0x00090] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at (wrapper remoting-invoke-with-check) System.Data.SQLite.SQLiteDataReader..ctor(System.Data.SQLite.SQLiteCommand,System.Data.CommandBehavior)
  at System.Data.SQLite.SQLiteCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x0000c] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLiteCommand.ExecuteScalar (System.Data.CommandBehavior behavior) [0x00006] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLiteCommand.ExecuteScalar () [0x00006] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at Marr.Data.QGen.InsertQueryBuilder`1[T].Execute () [0x00046] in C:\projects\radarr-usby1\src\Marr.Data\QGen\InsertQueryBuilder.cs:140 
  at Marr.Data.DataMapper.Insert[T] (T entity) [0x0005d] in C:\projects\radarr-usby1\src\Marr.Data\DataMapper.cs:728 
  at NzbDrone.Core.Datastore.BasicRepository`1[TModel].Insert (TModel model) [0x0002d] in C:\projects\radarr-usby1\src\NzbDrone.Core\Datastore\BasicRepository.cs:114 
  at NzbDrone.Core.MediaFiles.MediaFileService.Add (NzbDrone.Core.MediaFiles.MovieFile movieFile) [0x00000] in C:\projects\radarr-usby1\src\NzbDrone.Core\MediaFiles\MediaFileService.cs:44 
  at NzbDrone.Core.MediaFiles.MovieImport.ImportApprovedMovie.Import (System.Collections.Generic.List`1[T] decisions, System.Boolean newDownload, NzbDrone.Core.Download.DownloadClientItem downloadClientItem, NzbDrone.Core.MediaFiles.MovieImport.ImportMode importMode) [0x0028f] in C:\projects\radarr-usby1\src\NzbDrone.Core\MediaFiles\MovieImport\ImportApprovedMovie.cs:117 
```

I always have this issue when the movie doesn't have a year in it so I'm fine to do it manually but I can't do it because of this bug. I had this problem 3 times. I tried move and copy with the same issue.

I don't know why it doesn't work as a user ;)

**Comment by mlennox22** [comment]
Having the same problem on Radarr Ver. 0.2.0.1217.  The file is moved, but doesn't appear in Radarr due to the exception being thrown.  Happens with many movies during a Manual Import from the Wanted tab.

`19-1-6 22:01:58.4|Trace|ManualImportService|Processing file 9 of 10
19-1-6 22:01:58.4|Trace|EventAggregator|Publishing CommandUpdatedEvent
19-1-6 22:01:58.4|Trace|EventAggregator|CommandUpdatedEvent -> CommandModule
19-1-6 22:01:58.4|Trace|EventAggregator|CommandUpdatedEvent <- CommandModule
19-1-6 22:01:58.4|Trace|Owin|SQLite warning (284): automatic index on MovieFiles(MovieId)

19-1-6 22:01:58.4|Trace|ConfigService|Using default config value for 'parsingleniency' defaultValue:'Strict'
19-1-6 22:01:58.4|Debug|Parser|Parsing string 'Twas The Night Before Christmas.avi'
19-1-6 22:01:58.4|Debug|Parser|Unable to parse Twas The Night Before Christmas.avi
19-1-6 22:01:58.4|Debug|ParsingService|Attempting to parse movie info using directory and file names. Twas The Night Before Christmas
19-1-6 22:01:58.4|Trace|ConfigService|Using default config value for 'parsingleniency' defaultValue:'Strict'
19-1-6 22:01:58.4|Debug|Parser|Parsing string 'Twas The Night Before Christmas Twas The Night Before Christmas.avi'
19-1-6 22:01:58.4|Debug|Parser|Unable to parse Twas The Night Before Christmas Twas The Night Before Christmas.avi
19-1-6 22:01:58.4|Debug|ParsingService|Attempting to parse movie info using directory name. Twas The Night Before Christmas
19-1-6 22:01:58.4|Trace|ConfigService|Using default config value for 'parsingleniency' defaultValue:'Strict'
19-1-6 22:01:58.4|Debug|Parser|Parsing string 'Twas The Night Before Christmas.avi'
19-1-6 22:01:58.4|Debug|Parser|Unable to parse Twas The Night Before Christmas.avi
19-1-6 22:01:58.4|Debug|VideoFileInfoReader|Getting media info from /data/Media/Movies_radarr/Twas The Night Before Christmas/Twas The Night Before Christmas.avi
19-1-6 22:01:58.5|Trace|MediaInfo|Read file offset 0-16384 (16384 bytes)
19-1-6 22:01:58.6|Trace|MediaInfo|Read file offset 166569808-168036904 (1467096 bytes)
19-1-6 22:01:58.7|Trace|MediaInfo|Read file offset 4108-987148 (983040 bytes)
19-1-6 22:01:58.7|Trace|MediaInfo|Read a total of 2466520 bytes (1.5%)
19-1-6 22:01:58.7|Debug|ImportApprovedMovie|Decisions: 1
19-1-6 22:01:58.7|Trace|UpgradeMediaFileService|Upgrading existing movie file.
19-1-6 22:01:58.7|Debug|MediaInfoFormatter|Formatting audio channels using 'AudioChannels', with a value of: '2'
19-1-6 22:01:58.7|Trace|Http|Req: 767 [GET] /api/queue?sort_by=movie&order=asc
19-1-6 22:01:58.7|Debug|MediaInfoFormatter|Formatting audio channels using 'AudioChannels', with a value of: '2'
19-1-6 22:01:58.7|Trace|Http|Res: 767 [GET] /api/queue?sort_by=movie&order=asc: 200.OK (4 ms)
19-1-6 22:01:58.7|Debug|Api|[GET] /api/queue?sort_by=movie&order=asc: 200.OK (4 ms)
19-1-6 22:01:58.7|Debug|DiskProvider|Setting permissions: 0777 on /data/Media/Movies/'Twas the Night Before Christmas (1974)
19-1-6 22:01:58.7|Trace|EventAggregator|Publishing MovieFolderCreatedEvent
19-1-6 22:01:58.7|Trace|EventAggregator|MovieFolderCreatedEvent -> ExtraService
19-1-6 22:01:58.8|Debug|DiskProvider|Setting permissions: 0777 on /data/Media/Movies/'Twas the Night Before Christmas (1974)/poster.jpg
19-1-6 22:01:58.8|Debug|DiskProvider|Setting permissions: 0777 on /data/Media/Movies/'Twas the Night Before Christmas (1974)/fanart.jpg
19-1-6 22:01:58.8|Trace|Owin|SQLite error (5): database is locked

19-1-6 22:01:58.9|Trace|EventAggregator|MovieFolderCreatedEvent <- ExtraService
19-1-6 22:01:58.9|Debug|MovieFileMovingService|Moving movie file: /data/Media/Movies_radarr/Twas The Night Before Christmas/Twas The Night Before Christmas.avi to /data/Media/Movies/'Twas the Night Before Christmas (1974)/'Twas the Night Before Christmas 1974 [MP3 2.0 XVID].avi
19-1-6 22:01:58.9|Debug|DiskTransferService|Move [/data/Media/Movies_radarr/Twas The Night Before Christmas/Twas The Night Before Christmas.avi] > [/data/Media/Movies/'Twas the Night Before Christmas (1974)/'Twas the Night Before Christmas 1974 [MP3 2.0 XVID].avi]
19-1-6 22:01:58.9|Trace|SymbLinkResolver|Checking path /data/Media/Movies/'Twas the Night Before Christmas (1974)/'Twas the Night Before Christmas 1974 [MP3 2.0 XVID].avi for symlink returned error ENOENT, assuming it's not a symlink.
19-1-6 22:01:58.9|Trace|DiskTransferService|Attempting to move hardlinked backup.
19-1-6 22:01:58.9|Trace|DiskTransferService|Hardlink move succeeded, deleting source.
19-1-6 22:01:58.9|Trace|DiskProviderBase|Deleting file: /data/Media/Movies_radarr/Twas The Night Before Christmas/Twas The Night Before Christmas.avi
19-1-6 22:01:58.9|Trace|ConfigService|Using default config value for 'filedate' defaultValue:'None'
19-1-6 22:01:58.9|Debug|DiskProvider|Setting permissions: 0777 on /data/Media/Movies/'Twas the Night Before Christmas (1974)/'Twas the Night Before Christmas 1974 [MP3 2.0 XVID].avi
19-1-6 22:01:58.9|Trace|Owin|SQLite error (1299): abort at 16 in [INSERT INTO [MovieFiles] ([MovieId],[RelativePath],[Size],[DateAdded],[SceneName],[ReleaseGroup],[Quality],[MediaInfo],[Edition]) VALUES (@MovieId,@RelativePath,@Size,@DateAdded,@SceneName,@Rel

19-1-6 22:01:58.9|Warn|ImportApprovedMovie|Couldn't import movie /data/Media/Movies_radarr/Twas The Night Before Christmas/Twas The Night Before Christmas.avi

[v0.2.0.1217] System.Data.SQLite.SQLiteException (0x80004005): constraint failed
NOT NULL constraint failed: MovieFiles.Quality
  at System.Data.SQLite.SQLite3.Reset (System.Data.SQLite.SQLiteStatement stmt) [0x00083] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLite3.Step (System.Data.SQLite.SQLiteStatement stmt) [0x0003c] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLiteDataReader.NextResult () [0x0016b] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLiteDataReader..ctor (System.Data.SQLite.SQLiteCommand cmd, System.Data.CommandBehavior behave) [0x00090] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at (wrapper remoting-invoke-with-check) System.Data.SQLite.SQLiteDataReader..ctor(System.Data.SQLite.SQLiteCommand,System.Data.CommandBehavior)
  at System.Data.SQLite.SQLiteCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x0000c] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLiteCommand.ExecuteScalar (System.Data.CommandBehavior behavior) [0x00006] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLiteCommand.ExecuteScalar () [0x00006] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at Marr.Data.QGen.InsertQueryBuilder`1[T].Execute () [0x00051] in <b03e0e9d0a304e08a2033591f2e0b772>:0 
  at Marr.Data.DataMapper.Insert[T] (T entity) [0x0005d] in <b03e0e9d0a304e08a2033591f2e0b772>:0 
  at NzbDrone.Core.Datastore.BasicRepository`1[TModel].Insert (TModel model) [0x00033] in <eb6d7ee69d274d0eb2c19e6d87dabfdc>:0 
  at NzbDrone.Core.MediaFiles.MediaFileService.Add (NzbDrone.Core.MediaFiles.MovieFile movieFile) [0x00000] in <eb6d7ee69d274d0eb2c19e6d87dabfdc>:0 
  at NzbDrone.Core.MediaFiles.MovieImport.ImportApprovedMovie.Import (System.Collections.Generic.List`1[T] decisions, System.Boolean newDownload, NzbDrone.Core.Download.DownloadClientItem downloadClientItem, NzbDrone.Core.MediaFiles.MovieImport.ImportMode importMode) [0x0028f] in <eb6d7ee69d274d0eb2c19e6d87dabfdc>:0 
`

**Comment by LordMike** [comment]
I'm seeing this too, in ` Ver. 0.2.0.1344`

> Couldn't import movie /downloads/in-progress/movies/A Good Day to Die Hard (2013)/A Good Day to Die Hard (2013) - [EXTENDED][BLURAY-1080P][DTS-ES 6.1][X264]-SPARKS.mkv: constraint failed
> NOT NULL constraint failed: MovieFiles.Quality

```
System.Data.SQLite.SQLiteException (0x80004005): constraint failed
NOT NULL constraint failed: MovieFiles.Quality
  at System.Data.SQLite.SQLite3.Reset (System.Data.SQLite.SQLiteStatement stmt) [0x00083] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLite3.Step (System.Data.SQLite.SQLiteStatement stmt) [0x0003c] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLiteDataReader.NextResult () [0x0016b] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLiteDataReader..ctor (System.Data.SQLite.SQLiteCommand cmd, System.Data.CommandBehavior behave) [0x00090] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at (wrapper remoting-invoke-with-check) System.Data.SQLite.SQLiteDataReader..ctor(System.Data.SQLite.SQLiteCommand,System.Data.CommandBehavior)
  at System.Data.SQLite.SQLiteCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x0000c] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLiteCommand.ExecuteScalar (System.Data.CommandBehavior behavior) [0x00006] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at System.Data.SQLite.SQLiteCommand.ExecuteScalar () [0x00006] in <61a20cde294d4a3eb43b9d9f6284613b>:0 
  at Marr.Data.QGen.InsertQueryBuilder`1[T].Execute () [0x00046] in C:\projects\radarr-usby1\src\Marr.Data\QGen\InsertQueryBuilder.cs:140 
  at Marr.Data.DataMapper.Insert[T] (T entity) [0x0005d] in C:\projects\radarr-usby1\src\Marr.Data\DataMapper.cs:728 
  at NzbDrone.Core.Datastore.BasicRepository`1[TModel].Insert (TModel model) [0x00034] in C:\projects\radarr-usby1\src\NzbDrone.Core\Datastore\BasicRepository.cs:133 
  at NzbDrone.Core.MediaFiles.MediaFileService.Add (NzbDrone.Core.MediaFiles.MovieFile movieFile) [0x00000] in C:\projects\radarr-usby1\src\NzbDrone.Core\MediaFiles\MediaFileService.cs:44 
  at NzbDrone.Core.MediaFiles.MovieImport.ImportApprovedMovie.Import (System.Collections.Generic.List`1[T] decisions, System.Boolean newDownload, NzbDrone.Core.Download.DownloadClientItem downloadClientItem, NzbDrone.Core.MediaFiles.MovieImport.ImportMode importMode) [0x0028f] in C:\projects\radarr-usby1\src\NzbDrone.Core\MediaFiles\MovieImport\ImportApprovedMovie.cs:117 
```

**Comment by LordMike** [comment]
It's still present in `0.2.0.1358`, btw.

**Comment by LordMike** [comment]
The error is here, https://github.com/Radarr/Radarr/blob/01ad015b1433ce792c24f019f701f3a8a59c4b2c/src/NzbDrone.Core/MediaFiles/MovieImport/ImportApprovedMovie.cs#L117


.. but the interesting is actually here. In this line, it is assumed that the Quality of `localMovie` could be null, so to sort it some defaults are used ... But no similar check is made later, in L84, where the Quality is moved from `localMovie` to `movieFile`. I'm assuming the constraint has changed - as in it was originally allowed to be null, but now it isn't..

https://github.com/Radarr/Radarr/blob/01ad015b1433ce792c24f019f701f3a8a59c4b2c/src/NzbDrone.Core/MediaFiles/MovieImport/ImportApprovedMovie.cs#L55

**Comment by LordMike** [comment]
Some things were also fixed in `v0.2.0.935 (2018-02-06)` ([changelog](https://github.com/Radarr/Radarr/blob/5faaffc9c6a48e5a8bf6c04b2ec48370af6bd1ab/CHANGELOG.md#v020935-2018-02-06)), by `Leonardo Galli` with relations to this issue:

* Fixed Rare case of null quality causing issues with manual import. [Leonardo Galli]
* Fixed Error when no quality was associated with a movie file. [Leonardo Galli]

**Comment by jonathanklingberg** [comment]
Issue still present in `v0.2.0.1450`
Did you find a workaround?

**Comment by LordMike** [comment]
I have not. I even booted up a fresh install, pulled the database schema to see if I'd missed a migration .. but no. The fresh and my db were identical. 

I'm encountering the issue a lot less now though. I fixed a few other logical issues (downloading to incorrect folders - longstanding misconfiguration), and it seems to work well now (movies arriving automatically).

**Comment by Den4X** [comment]
ouldn't import movie \\MyPath\Goodfellas\Goodfellas.remux.mkv.mkv: constraint failed
NOT NULL constraint failed: MovieFiles.Quality

System.Data.SQLite.SQLiteException (0x80004005): constraint failed
NOT NULL constraint failed: MovieFiles.Quality
   в System.Data.SQLite.SQLite3.Reset(SQLiteStatement stmt)
   в System.Data.SQLite.SQLite3.Step(SQLiteStatement stmt)
   в System.Data.SQLite.SQLiteDataReader.NextResult()
   в System.Data.SQLite.SQLiteDataReader..ctor(SQLiteCommand cmd, CommandBehavior behave)
   в System.Data.SQLite.SQLiteCommand.ExecuteReader(CommandBehavior behavior)
   в System.Data.SQLite.SQLiteCommand.ExecuteScalar(CommandBehavior behavior)
   в System.Data.SQLite.SQLiteCommand.ExecuteScalar()
   в Marr.Data.QGen.InsertQueryBuilder`1.Execute() в C:\projects\radarr-usby1\src\Marr.Data\QGen\InsertQueryBuilder.cs:строка 138
   в Marr.Data.DataMapper.Insert[T](T entity) в C:\projects\radarr-usby1\src\Marr.Data\DataMapper.cs:строка 728
   в NzbDrone.Core.Datastore.BasicRepository`1.Insert(TModel model) в C:\projects\radarr-usby1\src\NzbDrone.Core\Datastore\BasicRepository.cs:строка 133
   в NzbDrone.Core.MediaFiles.MediaFileService.Add(MovieFile movieFile) в C:\projects\radarr-usby1\src\NzbDrone.Core\MediaFiles\MediaFileService.cs:строка 44
   в NzbDrone.Core.MediaFiles.MovieImport.ImportApprovedMovie.Import(List`1 decisions, Boolean newDownload, DownloadClientItem downloadClientItem, ImportMode importMode) в C:\projects\radarr-usby1\src\NzbDrone.Core\MediaFiles\MovieImport\ImportApprovedMovie.cs:строка 117

**Comment by Den4X** [comment]
maybe filename with double .mkv cause it?

**Comment by LordMike** [comment]
I don't think so.. you can test by renaming it, and then doing a manual import.

I assume the issue is that whatever logic is used to detect quality (1080p, 2160p etc.), has a condition where it ends up in NULL. I found in the code that it previously looks like they supported NULL's - but now they don't. All the code wasn't updated though, so there are code paths where Quality can end up in NULL.

**Comment by jonathanklingberg** [comment]
Looks like more or less my entire movie library is marked red with a note saying "Unable to process file" on the Wanted -> Manual Import page.
Radarr is unable to match the file with the right movie and I have to manually match the files with the right movie. After matching a file with a movie and click import I get this "NOT NULL constraint failed: MovieFiles.Quality" error.
I think all movies have already been added to the radarr library once so this is more of an annoyance than a real problem, but I'm not 100% sure.

**Comment by pish180** [comment]
I'm having the same issue.  Guess it has not been fixed yet. 

I have tried a few things.  
I moved all the files out of their folders into a new one and then the error changed. 

Now I get: 
Unable to apply permissions to: /downloads/Completed/movies/Anna-2019: Unknown user: brian

I was trying to MOVE the files and it apparently was not able to.  I went into the filing system and they files were owned by nobody (likely to do SAB not setting the permissions correctly). 

Then I try it again and I get: 
NzbDrone.Mono.Disk.LinuxPermissionsException: Error setting file permissions: EPERM

Long story short:
I had to move the files to a different location and then delete the movies from the Radarr database and then re-add them.  Once the movies were re-added I was able to import them.   I'm not sure if if this was because my image was shutdown improperly and the database was left corrupt or something, but deleting the movies seemed to clear it up.  now the Mover is properly moving the files to where they belong.  Its a bit of a PITA to have to delete all the movies from the DB but at least it is working.
